### PR TITLE
Custom Optimizer and LR schedulers

### PR DESCRIPTION
**Pull request by wistuba**
Optimizer and LR scheduler can be defined via Renate config file.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/renate/cli/parsing_functions.py [hunk 1]
```diff
@@ -348,56 +340,46 @@ def _add_hyperparameter_arguments(arguments: Dict[str, Dict[str, Any]]) -> None:
...
     parse_by_updater[updater](arguments)
-    _add_optimizer_arguments(arguments)
+    _add_optimizer_arguments(arguments, add_optimizer_args)
     for value in arguments.values():
         if "argument_group" not in value:
             value["argument_group"] = HYPERPARAMETER_ARGS_GROUP
 
 
-def _add_optimizer_arguments(arguments: Dict[str, Dict[str, Any]]) -> None:
+def _add_optimizer_arguments(
```
Comment (prabhuteja12) [inline src/renate/cli/parsing_functions.py:349]: What does the new boolean flag do?

File: src/renate/updaters/avalanche/model_updater.py [hunk 2]
```diff
@@ -61,10 +62,21 @@ def _load_learner(
...
         )
-        optimizer, scheduler = self._dummy_learner.configure_optimizers()
-        optimizer, scheduler = optimizer[0], scheduler[0]
-        lr_scheduler_plugin = LRSchedulerPlugin(scheduler=scheduler)
-        plugins = [lr_scheduler_plugin]
+        plugins = []
+        lr_scheduler_plugin = None
+        optimizers_scheduler = self._dummy_learner.configure_optimizers()
+        if isinstance(optimizers_scheduler, tuple):
+            optimizer, scheduler_config = optimizers_scheduler
```
Comment (prabhuteja12) [inline src/renate/updaters/avalanche/model_updater.py:69]: What do these two lines do?

**Review by prabhuteja12 — CHANGES_REQUESTED** [review]
The self.save_hyperparameters must now include the optimizer object.

File: test/conftest.py [hunk 3]
```diff
@@ -421,6 +290,8 @@ def get_simple_updater(
     metric=None,
     deterministic_trainer=False,
 ):
+    if learner_kwargs is None:
```
Comment (prabhuteja12) [inline test/conftest.py:293]: Why this form?

File: src/renate/updaters/avalanche/model_updater.py [hunk 4]
```diff
@@ -224,15 +238,11 @@ def __init__(
...
         memory_size: int,
         memory_batch_size: int = defaults.BATCH_SIZE,
-        optimizer: defaults.SUPPORTED_OPTIMIZERS_TYPE = defaults.OPTIMIZER,
-        learning_rate: float = defaults.LEARNING_RATE,
-        learning_rate_scheduler: defaults.SUPPORTED_LEARNING_RATE_SCHEDULERS_TYPE = defaults.LEARNING_RATE_SCHEDULER,  # noqa: E501
-        learning_rate_scheduler_gamma: float = defaults.LEARNING_RATE_SCHEDULER_GAMMA,
-        learning_rate_scheduler_step_size: int = defaults.LEARNING_RATE_SCHEDULER_STEP_SIZE,
-        momentum: float = defaults.MOMENTUM,
-        weight_decay: float = defaults.WEIGHT_DECAY,
+        learning_rate_scheduler: Optional[partial] = None,
```
Comment (prabhuteja12) [inline src/renate/updaters/avalanche/model_updater.py:244]: Type Callable?

File: src/renate/updaters/avalanche/model_updater.py (hunk 2 above)
Comment (wistuba) [reply src/renate/updaters/avalanche/model_updater.py:69]: Return value can either be Tuple[Optimizer, Scheduler] or Optimizer. We are checking for this case and conditionally add a LRCallback.

File: src/renate/updaters/avalanche/model_updater.py (hunk 4 above)
Comment (wistuba) [reply src/renate/updaters/avalanche/model_updater.py:244]: changed it

**Comment by wistuba** [comment]
> The self.save_hyperparameters must now include the optimizer object.

I've also ignored the lr_scheduler

File: test/conftest.py (hunk 3 above)
Comment (prabhuteja12) [reply test/conftest.py:293]: Yes. Is there a test that populates different `learner_kwargs`

File: doc/getting_started/how_to_run_training.rst [hunk 5]
```diff
@@ -47,6 +47,11 @@ instantiating the method you selected. See :doc:`supported_algorithms` for more
 
 
 
+..  note::
+    If you have defined the ``optimizer_fn`` function in your Renate config, do not pass values for the keys
```
Comment (prabhuteja12) [inline doc/getting_started/how_to_run_training.rst:51]: Is there a check for this in the code?

Comment (wistuba) [reply doc/getting_started/how_to_run_training.rst:51]: Yes. `run_training_job.py` will not add these params in the `argparse` code and therefore the script will fail if these arguments are passed but not expected.

File: test/conftest.py (hunk 3 above)
Comment (wistuba) [reply test/conftest.py:293]: yes, e.g., test_joint.py L27. This change is not changing the behavior but rather avoiding the unwanted behavior when using dicts and the likes as defaults in functions.

**Review by prabhuteja12 — APPROVED** [review]
